frontend/dockerfile/docs/reference.md

```diff
@@ -1372,100 +1423,133 @@ The available `[OPTIONS]` are:
+> changed. This includes invalidating the cache for `RUN` instructions.
+>
+> See [Dockerfile Best Practices guide – Leverage build cache](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#leverage-build-cache)
+> for more information.
 
-Each `<src>` may contain wildcards and matching will be done using Go's
-[filepath.Match](https://golang.org/pkg/path/filepath#Match) rules. For example:
+### Source
 
-To add all files in the root of the build context starting with "hom":
+You can specify multiple source files or directories with `COPY`. The last
+argument must always be the destination. For example, to copy two files,
+`file1.txt` and `file2.txt`, from the build context to `/usr/src/things/` in
+the build container:
 
 ```dockerfile
-COPY hom* /mydir/
+COPY file1.txt file2.txt /usr/src/things/
 ```
 
-In the following example, `?` is a single-character wildcard, matching e.g. "home.txt".
+If you specify multiple source files, either directly or using a wildcard, then
+the destination must be a directory (must end with a slash `/`).
+
+`COPY` accepts a flag `--from=<name>` that lets you specify the source location
+to be a build stage, context, or image. The following example copies files from
+a stage named `build`:
 
 ```dockerfile
-COPY hom?.txt /mydir/
+FROM golang AS build
+WORKDIR /app
+RUN --mount=type=bind,target=. go build -o /myapp ./cmd
+
+COPY --from=build /myapp /usr/bin/
 ```
 
-The `<dest>` is an absolute path, or a path relative to `WORKDIR`, into which
-the source will be copied inside the destination container.
+For more information about copying from named sources, see the
+[`--from` flag](#copy---from).
+
+#### Copying from the build context
+
+When copying source files from the build context, their paths are interpreted as
+relative to the root of the context. If you specify a relative path leading
+outside of the build context, such as `COPY ../something /something`, parent
+directory paths are stripped out automatically. The effective source path in
+this example becomes `COPY something /something`.
+
+If the source is a directory, the contents of the directory are copied,
+including filesystem metadata. The directory itself isn't copied, only its
+contents. If it contains subdirectories, these are also copied, and merged with
+any existing directories at the destination. Any conflicts are resolved in
+favor of the content being added, on a file-by-file basis.
+
+If the source is a file, the file and its metadata are copied to the
+destination. File permissions are preserved.
+
+If you pass a Dockerfile through stdin to the build (`docker build - <
+Dockerfile`), there is no build context. In this case, you can only use the
+`COPY` instruction to copy files from other stages, named contexts, or images,
+using the [`--from` flag](#copy---from). You can also pass a compressed archive
+through stdin: (`docker build - < archive.tar.gz`), the Dockerfile at the root
```

Comment:
Minor nit that the archive doesn't have to be compressed, but perhaps that's not important here? (you can see that by doing something like `gunzip < archive.tar.gz | docker build -`)

Reply:
good catch 👍🏻 I removed the "compressed" mention.